unliftio/src/UnliftIO/IO/File/Posix.hs

```diff
@@ -580,3 +582,13 @@ withBinaryFileAtomic filePath iomode action =
       liftIO $ atomicTempFileRename Nothing mFileMode eTmpFile filePath
       pure res
 
+
+-- | See `renameFileDurable`
+renameFileDurable ::
+     MonadIO m => FilePath -> FilePath -> m ()
+renameFileDurable oldName newName =
+     liftIO $ withDirectory (takeDirectory oldName) $ \oldDirFd ->
+                withDirectory (takeDirectory newName) $ \newDirFd -> do
+                  renameFile oldName newName
+                  fsyncDirectoryFd "renameFileDurable" newDirFd
+                  fsyncDirectoryFd "renameFileDurable" oldDirFd
```

Comment:
Possible optimisation:

For the common case that `oldName` and `newName` are in the same directory, do only 1 `withDirectory` and do only 1 `fsync()`.